File: spring-web-java/src/test/java/spring/web/java/application/service/RequestLockServiceTest.java
package spring.web.java.application.service;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import spring.web.java.domain.repository.RequestLockRepository;
import spring.web.java.interfaces.exception.DomainException;

@ExtendWith(MockitoExtension.class)
class RequestLockServiceTest {

	@InjectMocks
	private RequestLockService requestLockService;

	@Mock
	private RequestLockRepository requestLockRepository;

	@Test
	@DisplayName("유효성 검사 기능은 데이터가 없을 경우 저장한다")
	void case1() {
		// Given
		Long memberId = 1L;
		String method = "GET";
		String uri = "/members";

		Mockito.when(requestLockRepository.setIfAbsent(memberId, method, uri)).thenReturn(true);

		// When
		requestLockService.validate(memberId, method, uri);

		// Then
		Mockito.verify(requestLockRepository, Mockito.times(1))
			.setIfAbsent(memberId, method, uri);
	}

	@Test
	@DisplayName("유효성 검사 기능은 데이터가 있을 경우 DomainException을 던진다")
	void case2() {
		// Given
		Long memberId = 1L;
		String method = "GET";
		String uri = "/members";

		Mockito.when(requestLockRepository.setIfAbsent(memberId, method, uri)).thenReturn(false);

		// When & Then
		Assertions.assertThatThrownBy(() -> requestLockService.validate(memberId, method, uri))
			.isInstanceOf(DomainException.class);

		Mockito.verify(requestLockRepository, Mockito.times(1))
			.setIfAbsent(memberId, method, uri);
	}
}

File: spring-web-kotlin/src/test/kotlin/spring/web/kotlin/application/service/RequestLockServiceTest.kt
package spring.web.kotlin.application.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import spring.web.kotlin.domain.repository.RequestLockRepository
import spring.web.kotlin.presentation.exception.DuplicateRequestException

class RequestLockServiceTest : DescribeSpec({
    val requestLockRepository = mockk<RequestLockRepository>()
    val requestLockService = RequestLockService(requestLockRepository)

    describe("validate 는") {
        val memberId = 1L
        val method = "GET"
        val uri = "/members"

        context("데이터가 없을 경우") {
            it("저장한다") {
                every { requestLockRepository.setIfAbsent(memberId, method, uri) } returns true

                requestLockService.validate(memberId, method, uri)

                verify(exactly = 1) { requestLockRepository.setIfAbsent(memberId, method, uri) }
            }
        }

        context("데이터가 있을 경우") {
            it("DuplicateRequestException 던진다") {
                every { requestLockRepository.setIfAbsent(memberId, method, uri) } returns false

                shouldThrow<DuplicateRequestException> { requestLockService.validate(memberId, method, uri) }
            }
        }
    }
})
